nautobot/dcim/graphql/types.py

```diff
@@ -61,61 +154,301 @@ def resolve_ip_addresses(self, args):
+            return peer
+        return None
+
+    def resolve_connected_power_feed(self, args):
+        peer = self.connected_endpoint
+        if peer and type(peer).__name__ == "PowerFeed":  # type built-in used to avoid class loading
+            return peer
+        return None
+
+    def resolve_connected_power_outlet(self, args):
+        peer = self.connected_endpoint
+        if peer and type(peer).__name__ == "PowerOutlet":  # type built-in used to avoid class loading
+            return peer
+        return None
+
+
+class RearPortType(gql_optimizer.OptimizedDjangoObjectType):
+    """Graphql Type Object for RearPort model."""
+
+    class Meta:
+        model = RearPort
+        filterset_class = RearPortFilterSet
+
+    cable_peer_circuit_termination = graphene.Field("nautobot.circuits.graphql.types.CircuitTerminationType")
+    cable_peer_console_port = graphene.Field("nautobot.dcim.graphql.types.ConsolePortType")
+    cable_peer_console_server_port = graphene.Field("nautobot.dcim.graphql.types.ConsoleServerPortType")
+    cable_peer_front_port = graphene.Field("nautobot.dcim.graphql.types.FrontPortType")
+    cable_peer_interface = graphene.Field("nautobot.dcim.graphql.types.InterfaceType")
+
+    def resolve_cable_peer_circuit_termination(self, args):
+        peer = self.get_cable_peer()
+        if type(peer).__name__ == "CircuitTermination":  # type built-in used to avoid class loading
+            return peer
+        return None
+
+    def resolve_cable_peer_console_port(self, args):
+        peer = self.get_cable_peer()
+        if type(peer).__name__ == "ConsolePort":  # type built-in used to avoid class loading
+            return peer
+        return None
+
+    def resolve_cable_peer_console_server_port(self, args):
+        peer = self.get_cable_peer()
+        if type(peer).__name__ == "ConsoleServerPort":  # type built-in used to avoid class loading
+            return peer
+        return None
+
+    def resolve_cable_peer_front_port(self, args):
+        peer = self.get_cable_peer()
+        if type(peer).__name__ == "FrontPort":  # type built-in used to avoid class loading
+            return peer
+        return None
+
+    def resolve_cable_peer_interface(self, args):
+        peer = self.get_cable_peer()
+        if type(peer).__name__ == "Interface":  # type built-in used to avoid class loading
+            return peer
+        return None
 
 
 class CableTerminationTypes(graphene.Union):
-    """GraphQL type for models that can terminate a Circuit."""
+    """GraphQL type for models that can be terminated on a cable."""
 
     class Meta:
         types = (
-            InterfaceType,
+            ConsolePortType,
             ConsoleServerPortType,
             CircuitTerminationType,
+            FrontPortType,
+            InterfaceType,
+            PowerFeedType,
+            PowerOutletType,
+            PowerPortType,
+            RearPortType,
         )
```

Comment:
Double-check this - this type is used for the `connected_endpoint` field on `PathEndpointMixin` - as such I think it's not the list of types that can terminate a cable, but actually the list of types that can terminate a CablePath? Might need to rename it to something like `CablePathEndpointTypes` and create a new, separate `CableTerminationTypes` union field... maybe then also add a generic `cable_peer` field to all appropriate models using this type?

Reply:
As I'm thinking about it, this probably actually needs to be moved to each individual model, and the `PathEndpointMixin` removed. The possible values for a `connected_endpoint` is the intersection of the available PathEndpoints (`Interface`, `ConsolePort`, `ConsoleServerPort`...etc) and the available endpoints for a given model defined by the [`COMPATABLE_TERMINATION_TYPES` global var](https://github.com/nautobot/nautobot/blob/bf9f46b074ff1d060afba17d762e06bfd320b9a5/nautobot/dcim/constants.py#L80). Because of this, I don't think we can use a mixin unless that mixin is somehow dynamically constructed to contain only the possible types for a given model.

Reply:
After playing with the code a little bit more, I decided to rename that to `PathEndpointTerminationTypes` and add an equivalent `CableTerminationTypes`. I'm wondering if this is working as we expect it to though. On the current version of Nautobot, I'm seeing the only available sub selection for `connected_endpoint` as `__typename`. This is on interface objects which do have a connected endpoint. The `__typename` does return the correct name of the type for the connected object

```graphql
query ($device: [String!], $interface: [String!]) {
  devices(name: $device) {
    name
    interfaces(name: $interface) {
      name
      id
      connected_endpoint {
        __typename
      }
    }
  }
}
```

returns
```graphql
{
  "data": {
    "devices": [
      {
        "name": "redacted",
        "interfaces": [
          {
            "name": "TenGigE0/0/0/29/0",
            "id": "5451c622-ed10-4597-96c9-ceeaf27ab14c",
            "connected_endpoint": {
              "__typename": "CircuitTerminationType"
            }
          }
        ]
      }
    ]
  }
}
```

Reply:
See https://github.com/nautobot/nautobot/pull/430#issuecomment-839779674 - when dealing with Union types, the query has to branch on the concrete type it contains in order to access the fields of that type:

![image](https://user-images.githubusercontent.com/5603551/159280279-d2204893-01c6-492b-bbe8-bb0489709866.png)